### DISCUSSION TITLE
Rollback a long workflow transaction in case of any failure. (reposted from slack channel)

### DISCUSSION THREAD

**Discussion by bantyK** [uber/cadence#4648] [Q&A] [answered]
My question is around compensation logic and rollback in case of failures in activities. 

Consider a scenario like this. I have a workflow which is listening for some signal. When that signal is received, I need to perform ‘n’ number of operations which include making some external API calls, making some DB updates and updating the workflow state. I would like to have  different activities to perform these tasks. Now lets say there is an error while doing any of these tasks, I would like to roll back to the previous state of my workflow, including DB rollbacks(if I have made any). What would be the best way to implement such logic inside my workflow?

I was looking at the compensation example but that example is handling failures in individual activities, in my case, I want to roll back potentially several activities at once. 

I am using Java SDK

**Comment by longquanzheng** [comment, marked as answer]
Some points from myself:
* You cannot split a DB transaction into different Cadence activities. It simply doesn't work -- because you can't serialize/deserialize the DB connection and pass around Workflow to activities
* There are different ways to compensate the activities. You can use the Saga pattern similar to this [sample](https://github.com/uber/cadence-java-samples/blob/2fbc84a72a529edcb23fa3d4caf914f0fb8339ec/src/main/java/com/uber/cadence/samples/bookingsaga/TripBookingWorkflowImpl.java#L48), or you can write your own exception handling logic to handle ActivityFailureException
* For the three types of activity work load, typically workflow state is easiest to rollback, cause it's fully controlled by the workflow code. API is also easy to do as long as it has dedicated support for reverting the operation. Usually a DB transaction is difficult to revert. So an idea is that make the activity for DB transaction as the last step in the workflow, so that you don't need to rollback it. When it fails, just need to rollback others. 
* Essentially, if you consider all activities in a Cadence workflow is a transaction, the guarantee is eventual consistency.

**Reply by bantyK** [reply]
Thanks, @longquanzheng for your help.
